doc/cookbook/current-route/CurrentRoute.lhs

```diff
@@ -0,0 +1,391 @@
+{-# LANGUAGE OverloadedStrings       #-}
+{-# LANGUAGE RankNTypes              #-}
+{-# LANGUAGE ScopedTypeVariables     #-}
+{-# LANGUAGE TypeFamilies            #-}
+{-# LANGUAGE TypeOperators           #-}
+{-# LANGUAGE UndecidableInstances    #-}
+{-# LANGUAGE UndecidableSuperClasses #-}
+{-# OPTIONS_GHC -Wno-orphans         #-}
+module Main where
+
+import Data.Maybe             (fromMaybe)
+import Control.Monad.IO.Class (liftIO)
+import System.Environment     (getArgs)
+import GHC.Generics           (to, from, M1 (..), K1 (..), (:*:) (..))
+
+-- for "unsafe" vault key creation
+import System.IO.Unsafe (unsafePerformIO)
+
+import qualified Data.ByteString.Char8    as BS8
+import qualified Data.Vault.Lazy          as V
+import qualified Network.Wai              as Wai
+import qualified Network.Wai.Handler.Warp as Warp
+
+import Servant
+import Servant.API.Generic 
+import Servant.Server.Generic
+import Servant.Server.Internal.RoutingApplication (passToServer)
+```
+
+## How to get current `Request` object or URL
+
+Passing current `WAI` `Request` to the handler is actually quite easy.
+This is "lazy" approach, we ask for "everything" in the request,
+and we have to be careful in the handler (e.g. we cannot touch `requestBody`).
+Also this "combinator" ties implementation to the `wai` server implementation,
+which is an implementation detail
+(nothing else in `servant-server` exposes `wai` internals, except of `Raw`).
+
+The idea is to make `Server (Wai.Request :> api) = Wai.Request -> Server api`.
+If we imagine for a second that we have such functionality in place,
+we can write, using `Servant.API.Generic` (see "Using generics" cookbook recipe):
+
+```haskell
+data Routes1 route = Routes1
+    { route11 :: route :- Wai.Request :> "route1" :> Get '[JSON] Int
+    , route12 :: route :- Wai.Request :> "route2" :> Capture "id" Int :> Get '[JSON] Int
+    }
+  deriving (Generic)
+
+routes1 :: Routes1 AsServer
+routes1 = Routes1
+    { route11 = \req -> liftIO $ do
+        let p = Wai.rawPathInfo req
+        BS8.putStrLn p
+        return (BS8.length p)
+    , route12 = \req i -> liftIO $ do
+        let p = Wai.rawPathInfo req
+        BS8.putStrLn p
+        return (succ i)
+    }
+
+app1 :: Application
+app1 = genericServe routes1
+```
+
+We define a `Routes1` data type, implement `Routes1 AsServer` value and turn it
+into the `wai`'s `Application`. However, to compile this example, we need an
+additional instance.  We use an *internal* `passToServer` combinator in the
+implementation of `route`.
+
+```haskell
+instance HasServer api ctx => HasServer (Wai.Request :> api) ctx where
+    type ServerT (Wai.Request :> api) m = Wai.Request -> ServerT api m
+    
+    hoistServerWithContext _ pc nt s =
+        hoistServerWithContext (Proxy :: Proxy api) pc nt . s
+
+    route _ ctx d = route (Proxy :: Proxy api) ctx $
+        passToServer d id
+```
+
+This solution is good quick fix, but there are arguably better ways.
```

Comment:
While it's obviously fine not to dive into deep explanations of how the internals work, perhaps a sentence or two + a link to `passToServer`'s haddocks would help?

If you don't feel inspired I can give it a shot.

Reply:
Hmm. While looking up internals, I was confused about non-granular enough split in `servant-server`. I'll do a PR for that for `servant-0.16` still, and then it will be clearer where to add docs.

E.g. there are no `Delayed` module, which makes finding it difficult.

Reply:
Yeah, Delayed deserves its module.